### ISSUE TITLE
Processing multiple subprojects

### ISSUE THREAD

**Issue by afrendeiro** [epigen/looper#18]
I'd like to propose a change in the usage of subprojects.

When using subprojects, only the selected one (with the `--sp` argument) will be used.
This means that you have to run a looper command for each subproject (sounds like no big deal, but imagine if you have >5).

I propose to either:
1. change the default to when not given any `--sp` argument to run all subprojects
2. implement passing a keyword to run all subprojects (e.g. `--sp all`).

*and* to allow the passing of several subprojects `--sp prj1 prj2 prj3`.

Similarly, `looper summarize` will not make a summary of all subprojects in same file (currently they're separate) - the summary table should reflect the subprojects selected.

**Comment by nsheff** [comment]
glad to see you've realized the power of subprojects! excellent suggestion.

I prefer option 2, because option 1 isn't how I've used them... I always define my main project as a real project, and the subprojects as addendums. Option 1 requires thinking about it differently, where even your main project is a subproject.

But option 2 is great and should not be hard to implement. It will require an outer loop outside the main looper calls that handles `--sp` when it's a list.

Can you clarify what you're looking for with`summarize`? You want summarize to put all into 1 file or not?

**Comment by afrendeiro** [comment]
I'm using subprojects for a lot of external data and in this case there isn't even any main project, but many subprojects (small and big).

Option 2 is fine with me.
For the summarize, yes, I'd like to have one file for whatever subprojects were selected, but it's a minor thing.

**Comment by afrendeiro** [comment]
I just had a look at how subprojects are implemented and how could we implement this. I think it might not be so easy (to do it right).

Currently a subproject updates the entries in the main config if I'm not mistaken:
```python
self.add_entries(self.config['subprojects'][subproject])
```
What I want would be more like:
```python
for subproject in subprojects:
    # would actually "add" (as in append) entries
    self.add_entries(self.config['subprojects'][subproject])
```
This of course has implications in for example conflicting fields for different subprojects.

We could of course implement this in a `looper run`-oriented way, by instantiating `Project()` several times for each subproject, but that's just another workaround and not something I want to do (it would defeat the purpose of having a project object). Maybe the initial subproject implementation was just not meant to do this. Let's maybe leave this for after the merge.

**Comment by nsheff** [comment]
In my original thinking, it doesn't make sense to initialize multiple subprojects within a single Project -- the subproject is just a way of tweaking a Project. So you would instantiate a different Project for each subproject.
It would be one Project for each subproject -- to me this doesn't defeat the purpose of the Project object. You're dealing with multiple projects.

I don't really yet see your vision -- why would you want to do it that way, as it would overwrite variables... you could only do that for some variables for which append makes sense.

In other words, what you're describing could only work for very minimal subprojects, like those that just define a different annotation sheet. Then you could just merge annotation sheets. But what if subprojects have different compute characteristics? Different genomes? Right now, subprojects implement all the power of Projects, which is incompatible with your suggestion. This would require a new concept.

But just instantiating a Project for every subproject can do both things.

**Comment by nsheff** [comment]
@afrendeiro did you want to comment further on this, is it still relevant, or can we close this issue?

**Comment by afrendeiro** [comment]
I guess it's okay if this doesn't get implemented - merging annotation sheets would probably be a much simpler thing to do
